[RFC] terminal: Handle loss of focus in event loop.

splinterofchaos opened the pull request:
While in a terminal and insert mode, if an aucmd or event caused loss of focus, nvim would stay in the terminal event loop causing an inconsistent view of internal state and/or segfault.

Remove the "term" argument from terminal_enter() as it only makes sense to call it with curbuf->terminal. Terminate the loop either when focus is lost or the terminal closes.

~~Implement redraw() in terms of the pointer...~~

fixes #2301

@tarruda, for `terminal_enter()`, I notice that it has only one call site, and that `process_deferred` is unconditionally true. Do or will we need this to be called with `process_deferred=false` at any point, or can we simplify this?

Commit:
terminal: Handle loss of focus in event loop.

While in a terminal and insert mode, if an event caused loss of focus,
nvim would stay in the terminal event loop causing an inconsistent view
of internal state and/or segfault.

Remove the "term" argument from terminal_enter() as it only makes sense
to call it with curbuf->terminal. Terminate the loop when switched to a
different buffer.

fixes #2301

splinterofchaos commented:
Running `build/bin/tty-test` shows that my logic is a little off, so moving this back to WIP. Also, some of the output I got through the travis test is a little nonsensical.

splinterofchaos commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -155,5 +157,42 @@ describe('terminal buffer', function()
+      {2: }                                                 |
+      {1:+++                                               }|
+                                                        |
+    ]])
```
One outstanding issue: event when I put `execute('set noruler')`, that `rows: 2, cols: 50` shows up, and if the `statusline` is longer than one character, all but the first get appended after the `cols: 50`.

Otherwise, this is basically the output I expect.

splinterofchaos commented:
Failed on the quick-build tests, but I'm afraid I don't understand how to see the error.

Commit:
term: after <C-\>, resume normal input loop

Pressing <C-\> and then a mouse click will insert the click into the
terminal as if a keyboard button had been pressed.

Keep track of whether the last input was <C-\> and only call
terminal_send_key() if the next input is a key press.

splinterofchaos commented:
Marking RFC because I got the tests working,

Also, I noticed that if the user typed `<C-\>` and then either a mouse click or lost focus, the mouse click would be sent as a keyboard press, so I added a commit that lets us keep listening for `<C-\><C-n>`, but allows mouse presses and `K_EVENT`s as well.

Commit:
term: use window col offset to calculate width

fixes #2317

splinterofchaos commented:
Added a commit for #2317 (miscalculated terminal width), though I'm not quite sure how to test it except manually.

philix commented on src/nvim/terminal.c:
```diff
@@ -353,8 +353,14 @@ void terminal_resize(Terminal *term, uint16_t width, uint16_t height)
+void terminal_enter(bool process_deferred)
 {
+  Terminal *term = curbuf->terminal;
+  assert(term && "should only be called when curbuf has a terminal");
```
What if you pass the buffer to `terminal_enter` instead of relying on the `curbuf` global? You can move the `assert` up on the call chain and handle the error (or totally avoid it) in a better way.

I see that the terminal also has a reference to the buffer. Maybe this is a sign that most the function should actually take the buffer as argument. @tarruda may have his reasons.

tarruda replied:
> I see that the terminal also has a reference to the buffer. Maybe this is a sign that most the function should actually take the buffer as argument.

Yes, a `Terminal` is always tied to a `buf_T`. And since `buf_T->terminal` is used to distinguish normal buffers from terminal buffers, I guess it makes sense to refactor most functions in this module to receive buffers, thus avoiding the need for cyclic references(In another PR, of course)

splinterofchaos replied:
I did it like this because, if I understand the event loop correctly, we have to exit when `curbuf->terminal != term` or `term->buf != curbuf`, so even if `terminal_enter()` takes a `buf_T` argument, we'd have to assert `buf == curbuf && buf->terminal`.

tarruda replied:
> I did it like this because, if I understand the event loop correctly, we have to exit when curbuf->terminal != term or term->buf != curbuf, so even if terminal_enter() takes a buf_T argument, we'd have to assert buf == curbuf && buf->terminal.

:+1: for `terminal_enter` this makes sense(little gain in keeping the parameter since we still depend on the global)

tarruda commented on src/nvim/terminal.c:
```diff
@@ -413,12 +413,22 @@ void terminal_enter(Terminal *term, bool process_deferred)
 
+      case Ctrl_N:
+        if (got_bs) {
+          goto end;
```
Won't this result in `<c-n><c-\><c-n>` exiting terminal focus?

Commit:
term: ensure term->buf is valid

The fallowing test (reduced), submitted by @mhinz may free term->buf,
leaving the pointer dangling.

```vim
let s:buf  = -1

function! s:exit_handler()
  execute 'bdelete!' s:buf
endfunction

vnew
let s:buf = bufnr('%')
let id = termopen('sleep 1', { 'on_exit': function('s:exit_handler') })

call s:test()
```

When the buffer is known to be closing, set term->buf to NULL, and check
buf_valid() in on_refresh().

Helped-by: Marco Hinz (@mhinz)

splinterofchaos commented:
I don't think I'll be adding anything more to this right now. The last commit, "term: ensure term->buf is valid", I don't think is a good permanent solution (see the TODO), but it at least prevents memory errors when the terminal's buffer is unexpectedly deleted, or the other way around.

philix commented:
LGTM although I'm not familiar with the terminal code.

tarruda commented:
:+1: LGTM. Great work @splinterofchaos, and thanks for taking some of the terminal bugs off my back :smile:

Pyrohh commented:
@splinterofchaos It's probably not a big deal, but next time you merge could you do it locally? Since these commits weren't very recently rebased, the merge commit is shown as the most recent commit, but you have to scroll down a bunch to see the rest of the commits. @justinmk presumably still uses these two shell functions to automate the process:

https://github.com/justinmk/config/blob/master/.bashrc#L222

splinterofchaos commented:
Sorry, yes I should have rebased. I'll make sure to do so in the future. Thanks.
